### PR TITLE
Quell nullability warnings in src

### PR DESCRIPTION
**Pull request by blairconrad**
Contributes to #1613

### CODE REVIEW

**Comment by blairconrad** [comment]
This could close issue 1613, or we could fold #1691 into that issue and fix it when I clean up the tests. I've started that work. Only 334 warnings to go!

File: src/FakeItEasy/Creation/ProxyGeneratorResult.cs [hunk 1]
```diff
@@ -66,23 +66,22 @@ public ProxyGeneratorResult(object generatedProxy)
...
 
-            this.ProxyWasSuccessfullyGenerated = true;
             this.GeneratedProxy = generatedProxy;
         }
 
         /// <summary>
         /// Gets a value indicating whether the proxy was successfully created.
         /// </summary>
-        public bool ProxyWasSuccessfullyGenerated { get; }
+        public bool ProxyWasSuccessfullyGenerated => this.GeneratedProxy is object;
```
Comment (thomaslevesque) [inline src/FakeItEasy/Creation/ProxyGeneratorResult.cs]: I thought about something the other day: we *could* use `is notnull` instead of `is object`! We'd just need a `using notnull = System.Object;` in every file... Of course, it could lead to weird things, such as `notnull.Equals(...)` or `new notnull()`, but I like the idea anyway ;)

File: src/FakeItEasy/Creation/ProxyOptions.cs [hunk 2]
```diff
@@ -25,7 +26,8 @@ internal class ProxyOptions : IProxyOptions
 
         public IEnumerable<Expression<Func<Attribute>>> Attributes => this.attributes;
 
-        public string Name { get; set; }
+        [DisallowNull]
```
Comment (thomaslevesque) [inline src/FakeItEasy/Creation/ProxyOptions.cs]: Is there any benefit to adding this attribute here? We never explicitly set it to null, but we could; it's internal anyway. Also we don't check if the assigned value is null, so the attribute basically a lie... Unless it's necessary to quell a warning, I'd be inclined to remove it.

File: src/FakeItEasy/Expressions/CallExpressionParser.cs [hunk 3]
```diff
@@ -11,19 +11,12 @@ internal class CallExpressionParser : ICallExpressionParser
     {
         public ParsedCallExpression Parse(LambdaExpression callExpression)
         {
-            var methodExpression = callExpression.Body as MethodCallExpression;
-            if (methodExpression is object)
+            return callExpression.Body switch
```
Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/CallExpressionParser.cs]: I saw what you did here! 👍

File: src/FakeItEasy/Expressions/CallExpressionParser.cs [hunk 4]
```diff
@@ -39,7 +32,7 @@ public ParsedCallExpression Parse(LambdaExpression callExpression, object fake)
         private static ParsedCallExpression ParseInvocationExpression(InvocationExpression expression)
         {
             var target = expression.Expression.Evaluate();
-            var method = target.GetType().GetMethod("Invoke");
+            var method = target!.GetType().GetMethod("Invoke");
```
Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/CallExpressionParser.cs]: Can we be sure `target` will never be null? It seems a pretty big assumption, when invoking an expression we know nothing about... Not sure what to do if it's null, though... I guess we'll just throw `NullReferenceException`. Which makes sense, since it would mean the user did something like this:
```csharp
Action action = null;
A.CallTo(() => action())...
```

File: src/FakeItEasy/Expressions/ParsedCallExpression.cs [hunk 5]
```diff
@@ -16,7 +16,7 @@ internal class ParsedCallExpression
         {
             this.CalledMethod = calledMethod;
             this.ArgumentsExpressions = argumentsExpressions;
-            this.callTarget = new Lazy<object>(() => callTargetExpression?.Evaluate());
+            this.callTarget = new Lazy<object>(() => callTargetExpression?.Evaluate() !);
```
Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/ParsedCallExpression.cs]: Two things:
- Why the `?` (I know, it was already there)?. `callTargetExpression` is of type `Expression` (not nullable), so it should never be null. And if it is, it's an error.
- How do we know `Evaluate` won't return null?

File: src/FakeItEasy/Expressions/CallExpressionParser.cs (hunk 4 above)
Comment (thomaslevesque) [reply src/FakeItEasy/Expressions/CallExpressionParser.cs]: Anyway, do we really need to evaluate the expression here? We only use `target` to get its type, but we could get the type from `expression.Expression.Type`. Just tried it, it seems to work.

File: src/FakeItEasy/ServiceLocator.cs [hunk 6]
```diff
@@ -18,20 +18,20 @@ static ServiceLocator()
         [DebuggerStepThrough]
         internal static T Resolve<T>() where T : class
         {
-            var service = Service<T>.Instance;
+            var service = Service<T>.Instance!;
```
Comment (thomaslevesque) [inline src/FakeItEasy/ServiceLocator.cs]: Why is this necessary? We check for null on the next line

File: src/FakeItEasy.Analyzer.CSharp/FakeItEasy.Analyzer.CSharp.csproj [hunk 7]
```diff
@@ -6,7 +6,7 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <FileAlignment>512</FileAlignment>
     <SignAssembly>true</SignAssembly>
-    <DefineConstants>$(DefineConstants);CSHARP</DefineConstants>
+    <DefineConstants>$(DefineConstants);CSHARP;LACKS_NULLABLE_ATTRIBUTES</DefineConstants>
```
Comment (thomaslevesque) [inline src/FakeItEasy.Analyzer.CSharp/FakeItEasy.Analyzer.CSharp.csproj]: Is this necessary? We don't need the `#if`s, since there's a single TFM... Or is it future-proofing if we target netstandard2.1 someday?

Comment (thomaslevesque) [reply src/FakeItEasy.Analyzer.CSharp/FakeItEasy.Analyzer.CSharp.csproj]: Ah, I understand. The included file has `#if`s anyway, so we need the symbol

**Comment by thomaslevesque** [comment]
@blairconrad The latest changes look good thanks! There's just one unresolved comment left. Maybe it's fine, but I think there might be cases where `callTargetExpression` will evaluate to null.

File: src/FakeItEasy/Core/FakeManager.cs [hunk 8]
```diff
@@ -101,7 +101,7 @@ public virtual IEnumerable<IFakeObjectCallRule> Rules
         internal string FakeObjectDisplayName =>
             string.IsNullOrEmpty(this.FakeObjectName)
                 ? "Faked " + this.FakeObjectType
-                : this.FakeObjectName;
+                : this.FakeObjectName!;
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/Core/FakeManager.cs]: No, it's fine, I think

File: src/FakeItEasy/Creation/DelegateProxies/DelegateProxyGenerator.cs [hunk 9]
```diff
@@ -147,7 +147,7 @@ bool IsAccessibleImpl(Type t)
             // Return the result if the return type is non-void
             if (!isVoid)
             {
-                bodyExpressions.Add(result);
+                bodyExpressions.Add(result!);
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/Creation/DelegateProxies/DelegateProxyGenerator.cs]: I'm not sure it makes it more readable... Especially the `Insert`s at the end, which are a bit confusing. Anyway, I commented because I was surprised, but I wasn't asking for a change

File: src/FakeItEasy/Creation/ProxyGeneratorResult.cs (hunk 1 above)
Comment (thomaslevesque) [reply src/FakeItEasy/Creation/ProxyGeneratorResult.cs]: That was the idea

File: src/FakeItEasy/Expressions/ParsedCallExpression.cs (hunk 5 above)
Comment (blairconrad) [reply src/FakeItEasy/Expressions/ParsedCallExpression.cs]: You're right, `callTargetExpression` could evaluate to `null`. In particular when clients try to configure a static method. We catch the nullness/staticness later, but for a while, we can have a null call target.
I've fixed up. The nullability propagated bit. In 4 places I resorted to the `!` operator. In 3 of them, it's fairly obvious that it's safe. In `PropertyExpressionHelper`, I added comments to clear it up a bit…

Longer term, we might want to consider refactoring the determination of whether a method can be intercepted, but I'm leery of mucking about too much in the middle of the nullability transformation...

Comment (thomaslevesque) [reply src/FakeItEasy/Expressions/ParsedCallExpression.cs]: > I'm leery of mucking about too much in the middle of the nullability transformation...

Totally agree!

File: src/FakeItEasy/Configuration/FakeConfigurationManager.cs [hunk 10]
```diff
@@ -76,7 +76,7 @@ public IPropertySetterAnyValueConfiguration<TValue> CallToSet<TValue>(Expression
                     this.CreateVoidArgumentValidationConfiguration(fake, newParsedSetterCallExpression));
         }
 
-        private static void GuardAgainstNonFake(object target)
+        private static void GuardAgainstNonFake(object? target)
```
Comment (thomaslevesque) [inline src/FakeItEasy/Configuration/FakeConfigurationManager.cs]: Too bad there isn't an attribute to say "this method will throw if the argument is null". This way the compiler will now that the argument can't be null after the call. R# has an attribute like that ([example](https://github.com/thomaslevesque/Linq.Extras/blob/528233fef751cdfefbec3bb10096b96c41cd4d9e/src/Linq.Extras/Internal/ArgumentExtensions.cs#L10)).

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque. Excellent comments!

**Comment by thomaslevesque** [comment]
Thanks @blairconrad!
